mteb/evaluation/evaluators/model_classes.py

```diff
@@ -0,0 +1,562 @@
+        )
+
+        if top_ranked is not None:
+            logger.info("Performing reranking on pre-ranked documents...")
+            result_heaps = self._rerank_documents(
+                query_ids=query_ids,
+                query_embeddings=query_embeddings,
+                corpus=corpus,
+                top_ranked=top_ranked,
+                top_k=top_k,
+                score_function=score_function,
+                task_name=task_name,
+                request_qid=request_qid,
+                return_sorted=return_sorted,
+            )
+        else:
+            logger.info("Performing full corpus search...")
+            result_heaps = self._full_corpus_search(
+                query_ids=query_ids,
+                query_embeddings=query_embeddings,
+                corpus=corpus,
+                top_k=top_k,
+                score_function=score_function,
+                task_name=task_name,
+                request_qid=request_qid,
+                return_sorted=return_sorted,
+            )
+
+        for qid in result_heaps:
+            for score, corpus_id in result_heaps[qid]:
+                self.results[qid][corpus_id] = score
+
+        return self.results
+
+    def _rerank_documents(
+        self,
+        query_ids: list[str],
+        query_embeddings: torch.Tensor,
+        corpus: dict[str, dict[str, str]],
+        top_ranked: dict[str, list[str]],
+        top_k: int,
+        score_function: str,
+        task_name: str,
+        request_qid: str | None = None,
+        return_sorted: bool = False,
+    ) -> dict[str, list[tuple[float, str]]]:
+        """Rerank documents for each query using top_ranked."""
+        # Determine device
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        logger.info(f"Using device: {device}")
+
+        # Move query embeddings to appropriate device
+        query_embeddings = torch.as_tensor(query_embeddings).to(device)
+
+        result_heaps = {qid: [] for qid in query_ids}
+
+        # Get unique document IDs across all queries
+        unique_doc_ids = list(
+            {
+                doc_id
+                for qid in query_ids
+                if qid in top_ranked
+                for doc_id in top_ranked[qid]
+            }
+        )
+
+        # Create mapping from unique doc IDs to their index in the embedding matrix
+        doc_id_to_idx = {doc_id: idx for idx, doc_id in enumerate(unique_doc_ids)}
+
+        # Encode unique documents only once
+        unique_docs = [corpus[doc_id] for doc_id in unique_doc_ids]
+        all_doc_embeddings = self.model.encode(
+            unique_docs,
+            task_name=task_name,
+            prompt_type=PromptType.passage,
+            request_qid=request_qid,
+            **self.encode_kwargs,
+        )
+
+        # Let's make sure we don't get the warnings for the tokenizer here via torch.compile
+        if hasattr(torch, "compile"):
```

Comment:
It might be better to make compile as a flag or to check the CUDA version, as shown in [their tutorial](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html), since it can slow things down on certain architectures.